flink/src/main/java/io/delta/flink/sink/internal/committer/DeltaGlobalCommitter.java

```diff
@@ -212,86 +220,142 @@ private String resolveAppId(List<DeltaGlobalCommittable> globalCommittables) {
+                    appId,
+                    globalCommittables,
+                    deltaLog);
 
             for (long checkpointId : committablesPerCheckpoint.keySet()) {
-                OptimisticTransaction transaction = deltaLog.startTransaction();
-                long lastCommittedVersion = transaction.txnVersion(appId);
-                if (checkpointId > lastCommittedVersion) {
-                    doCommit(
-                        transaction,
-                        committablesPerCheckpoint.get(checkpointId),
-                        deltaLog.tableExists());
-                } else {
-                    LOG.info(String.format(
-                        "Skipping already committed transaction (appId='%s', checkpointId='%s')",
-                        appId, checkpointId));
-                }
+                doCommit(
+                    deltaLog.startTransaction(),
+                    committablesPerCheckpoint.get(checkpointId),
+                    deltaLog.tableExists());
             }
         }
+
+        this.firstCommit = false;
         return Collections.emptyList();
     }
 
+    private SortedMap<Long, List<CheckpointData>> getCommittablesPerCheckpoint(
+            String appId,
+            List<DeltaGlobalCommittable> globalCommittables,
+            DeltaLog deltaLog) {
+
+        OptimisticTransaction transaction = deltaLog.startTransaction();
+        long tableVersion = transaction.txnVersion(appId);
+
+        // processing recovery, deduplication on recovered committables.
+        if (this.firstCommit && tableVersion >= 0) {
+            LOG.info(
+                "Processing what it seems like, a first commit. This can be first commit ever for "
+                    + "this job or first commit after recovery.");
+
+            Map<String, CheckpointData> filePathToActionMap = new HashMap<>();
+
+            for (DeltaGlobalCommittable globalCommittable : globalCommittables) {
+                for (DeltaCommittable committable : globalCommittable.getDeltaCommittables()) {
+                    AddFile addFile = committable.getDeltaPendingFile().toAddFile();
+                    filePathToActionMap.put(
+                        addFile.getPath(),
+                        new CheckpointData(committable, addFile)
+                    );
+                }
+            }
+
+            Iterator<VersionLog> changes = deltaLog.getChanges(tableVersion, true);
+
+            StringJoiner duplicatedFiles = new StringJoiner(", ");
+            while (changes.hasNext()) {
+                VersionLog versionLog = changes.next();
+                try (CloseableIterator<Action> actionsIterator = versionLog.getActionsIterator()) {
+                    actionsIterator.forEachRemaining(action -> {
+                        if (action instanceof AddFile) {
+                            CheckpointData remove =
+                                filePathToActionMap.remove(((AddFile) action).getPath());
```

Comment:
can we double check that the `path` in this addfile (from the delta log) and the addfiles in the hashmap (from the committables) use the same absolute/relative/canonical format?

e.g. if we put `/absolute/table/path/_delta_log/00000.json` into the hashmap but the delta log returns to us `00000.json` then these lookups will always fial.

Reply:
2 questions for you @scottsand-db 
1. We are consider only by data that could be committed by this connector right?
2. delta log contains paths that were added by AddFile actions and pats are not beaing adjusted.
meaning whatever path was set on the AddFile action and committed to the Delta, will be added to the log without any change.

If answers for both questions are YES, then I think we don't need to perform a second lookup.
At the beginning I'm converting all committables to AddFile actions like [this](https://github.com/kristoffSC/connectors/blob/a913e5d8d4790060f009e6ab8fd6f0ae81a4ee39/flink/src/main/java/io/delta/flink/sink/internal/committer/DeltaGlobalCommitter.java#:~:text=AddFile%20addFile%20%3D%20committable.getDeltaPendingFile().toAddFile())

This method resolves file path taking under account partition definition. 

The same AddFile actions are committed to the delta log.

Let me know if my assumptions are incorrect.

Reply:
1. yes
2. delta-standalone will relativize paths https://github.com/delta-io/connectors/blob/master/standalone/src/main/scala/io/delta/standalone/internal/OptimisticTransactionImpl.scala#L245

Are we able to add to add a test case for both relative and absolute paths?